### PR TITLE
Revisit iOS interop API

### PR DESCRIPTION
**Pull request by elijah-semyonov**
## Changes

Revisit iOS interop API to build upon new implementation of `InteropView`. 
Main functional additions are:
Support of `onReset`, fine-grain touches strategy control (cooperative with explicit time delay, eager, none)ю

Fixes [5719](https://youtrack.jetbrains.com/issue/CMP-5719/iOS-revisit-UIKit-interop-API), [5876](https://youtrack.jetbrains.com/issue/CMP-5876/iOS-expose-API-for-selecting-strategy-of-touch-forwarding)

## Testing
TBD, new/updated demo pages.

This should be tested by QA

### Breaking changes - iOS
- `UIKitView` and `UIKitViewController` in `package androidx.compose.ui.interop` are deprecated. New API are mentioned in deprecation message. Deprecated invocations should work fine unless custom `onResize` is used, it is disallowed now and will print a warning.

### Features - iOS
- New `UIKitView` and `UIKitViewController` API in `package androidx.compose.ui.viewinterop`. Support of `onReset` to reuse the interop composable emitted node and avoid excessive native views reallocations, fine-grain touches strategy control (cooperative with explicit time delay, non-cooperative where no touches are received by Compose, ignoring touches).

### CODE REVIEW

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt]: Restore the doc and add `replaceWith` as separate parameter instead of writing it in the message

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt]: It seems like a breaking change if it was used previously

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropCallbacks.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropCallbacks.kt]: Please add KDoc for parameters, it's a public API

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt]: Nitpick: just regular `if`? double lambda wrapping seems weird

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt]: It seems semantic of this is simpler than name `isVisible` probably?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: `data` class cannot be public API. Just define equals/hashCode manually

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: Why are getters internal?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitView.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitView.uikit.kt]:
```suggestion
    modifier: Modifier = Modifier,
```

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitViewController.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitViewController.uikit.kt]:
```suggestion
    modifier: Modifier = Modifier,
```

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropCallbacks.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropCallbacks.kt]: `UIKitInteropListener`?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: Not sure that we want this as part of a public API

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt]: Yes, but it didn't make sense because it basically required user to set a proper frame of interop view, which was already laid out by compose.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: What would you like to do here?
I think we'd still expand this constructor/argument types, so want to keep at least a stable part without having a constructor overloads combinatory explosion

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropCallbacks.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropCallbacks.kt]: Sure, sounds better

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: I mean this particular default value. Let's make it internal for now (still can be used as default arg)

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt
Comment (MatkovIvan) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt]: It doesn't matter. It can be used in libraries that the user doesn't control, so it will become a blocker for Compose update

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt]: Done

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/interop/UIKitView.uikit.kt]: When did we stabilise this API in the first place?
I mean, I just don't want to carry a trail of compatibility code on API, that was experimental due to target being experimental.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt]: Yes, refactored it.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt]: Yeah, it's a side effect of my experimentation with reassignable callbacks/properties and callbacks being `var`. Changed it.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: Because it's only needed for us.

Comment (MatkovIvan) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: I'm not so sure. This is kind of an object where you just wrap a few booleans/enums. The expectation here is full creating + reading for the ability to create a new instance based on the some values from the existing one

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: Isn't it a JVM-only story? I don't think we support binary compatibility for Native. And either way we don't in fact even need `equals/hashCode`. Afaik `@Immutable` is already our guarantee on semantical immutability, so default `equals` with identity comparison is enough.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: Okay, let's make it public

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: https://kotlinlang.org/docs/api-guidelines-backward-compatibility.html#avoid-using-data-classes-in-your-api

Comment (MatkovIvan) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: I think we need to follow common rules here and put equals/hashCode explicitly here

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: Made it internal

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.kt]: Let's explicitly mark it as `@ExperimentalComposeUiApi`

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.kt]:
```suggestion
            const val DefaultDelayMillis = 150
```

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropListener.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropListener.kt]: Is this the final shape of this API? public-stable is intended here?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]: Unresolved

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt:59]: Please generate equals/hashCode

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt]:
```suggestion
    internal companion object {
```

**Review by MatkovIvan — COMMENTED** [review]
Also, such new APIs require a second reviewer

File: compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/UIKitInteropExample.kt
Comment (SebastianAigner) [inline compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/UIKitInteropExample.kt]: Wouldn't this listener object be recreated on each recomposition?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.kt]: Done

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropListener.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropListener.kt]: Actually, I've requested some insight from our dev advocates, so it could be not

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt:59]: Why? `@Immutable` implies reference comparisons.

File: compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/UIKitInteropExample.kt
Comment (elijah-semyonov) [reply compose/mpp/demo/src/uikitMain/kotlin/androidx/compose/mpp/demo/UIKitInteropExample.kt]: And the worst thing is that it won't be reapplied. Revisiting that part again.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropProperties.uikit.kt:59]: Done just in case.

**Comment by elijah-semyonov** [comment]
I rolled back additional APIs, only `onReset` is included.

**Comment by Gaubee** [comment]
Hope you can review and merge this PR as soon as possible. The change in touch strategy is having a big impact on our project, and we really need to push out a version that behaves correctly for our clients.

Thanks! 😊

**Comment by elijah-semyonov** [comment]
> Hope you can review and merge this PR as soon as possible. The change in touch strategy is having a big impact on our project, and we really need to push out a version that behaves correctly for our clients.
> 
> Thanks! 😊

You mean new touches strategy brought in recent alpha?

**Comment by kingsword09** [comment]
> > Hope you can review and merge this PR as soon as possible. The change in touch strategy is having a big impact on our project, and we really need to push out a version that behaves correctly for our clients.
> > Thanks! 😊
> 
> You mean new touches strategy brought in recent alpha?

Yes, in youtrack is [this](https://youtrack.jetbrains.com/issue/CMP-5705/ComposeUIViewController-Needs-to-Support-Hit-Through#focus=Comments-27-10239746.0-0).

**Comment by elijah-semyonov** [comment]
No worries, it's in its presumably final state, so it will be merged after reviewed.
But please, don't use alpha and dev builds for production.

**Comment by Gaubee** [comment]
> No worries, it's in its presumably final state, so it will be merged after reviewed.
> But please, don't use alpha and dev builds for production.

Understand, but because we are happy to conduct active exploration, we will provide customers with some experimental versions.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/viewinterop/InteropViewHolder.skiko.kt
Comment (ASalavei) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/viewinterop/InteropViewHolder.skiko.kt:141]: comments?

Comment (elijah-semyonov) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/viewinterop/InteropViewHolder.skiko.kt:141]: It's a reminder for investigation after upstream merge, ask @MatkovIvan for more details.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt:177]: `CGRectIntersectsRect(cgRect, group.bounds)` ?

Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt:158]: `CGRectIntersectsRec(cgRect, group.bounds)` ?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.uikit.kt
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.uikit.kt]: I would suggest to rename to `delayMs` or use specific type to represent duration on the platform.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt:177]: I considered using it, but there is a subtle difference:
[CGRectIntersectsRect](https://developer.apple.com/documentation/coregraphics/1454747-cgrectintersectsrect?language=objc) returns `true` if it's a `null` rectangle (absence of intersection).

It means it could still be `empty` rectangle, with either dimension equal to 0, which is still `not visible`.
[CGRectIsEmpty](https://developer.apple.com/documentation/coregraphics/1454917-cgrectisempty) includes both `null` and `empty` and this is exactly what we need.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropInteractionMode.uikit.kt]: Done

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt:177]: Deleted the duplicating review comment below (GitHub hiccup?)

Comment (ASalavei) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt:177]: Ok, understood

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/InteractionUIView.uikit.kt
Comment (mazunin-v-jb) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/InteractionUIView.uikit.kt]: just a tiny nitpick: as it was renamed to delayMs [here](https://github.com/JetBrains/compose-multiplatform-core/pull/1494#discussion_r1724902731), may be changing this here would be also nice

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/InteractionUIView.uikit.kt]: Makes sense indeed.

**Review by mazunin-v-jb — APPROVED** [review]

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/viewinterop/InteropPlatformDetails.skiko.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/viewinterop/InteropPlatformDetails.skiko.kt]: ~Why isn't it a part of `InteropContainer`? Those 2 booleans don't seem as platform-specific and can be mentioned in the common interface as parameters in this function.~

or. it might be converted to something like

```kt
internal fun interface InteropViewModifierFactory {
    fun create(holder: InteropViewHolder): Modifier
}
```

and probably make some default implementation(s) in shared code (outside of this PR).

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/viewinterop/UIKitInteropElementHolder.uikit.kt]: Resolving the previous comment should remove this cast, I guess

**Review by MatkovIvan — COMMENTED** [review]
I like the current shape of it, thanks!
(aside of `InteropPlatformDetails` in shared code 🙃 )

Also, could you please update PR description/release notes (no breaking changes here)

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/viewinterop/InteropPlatformDetails.skiko.kt
Comment (elijah-semyonov) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/viewinterop/InteropPlatformDetails.skiko.kt]: Native platform accessibility resolution is absolutely platform specific, touch interaction mode is also platform specific.
The point of `InteropPlatformDetails` is an ability to change something in the implementation of specific `InteropViewHolder` when it's reused. It can cover basically anything what happens here (in case of iOS - not only change of a modifier, but also - reconfiguration of `InteropWrappingView` which serves as a `group`. In future it could be used for expanding gesture recogniser failure requirements behavior and forwarding it to the user.

**Review by MatkovIvan — APPROVED** [review]

**Comment by vickyleu** [comment]
[1.7.0-dev1783](https://github.com/JetBrains/compose-multiplatform/releases/tag/v1.7.0-dev1783) 
In a full-screen WebView, to ensure smooth scrolling, you must set
```kotlin
properties = UIKitInteropProperties(
                    interactionMode= UIKitInteropInteractionMode.Cooperative(
                        delayMillis = 1,
                    ),
                    isNativeAccessibilityEnabled = false
)
```
I still don't quite understand the purpose of delaying touch events. Why is delay necessary? In most cases, delayMillis doesn't make much sense, as in scrollable scenarios, no one would hold down and wait for the page to become scrollable.

**Comment by Gaubee** [comment]
> [1.7.0-dev1783](https://github.com/JetBrains/compose-multiplatform/releases/tag/v1.7.0-dev1783) In a full-screen WebView, to ensure smooth scrolling, you must set
> 
> ```kotlin
> properties = UIKitInteropProperties(
>                     interactionMode= UIKitInteropInteractionMode.Cooperative(
>                         delayMillis = 1,
>                     ),
>                     isNativeAccessibilityEnabled = false
> )
> ```
> 
> I still don't quite understand the purpose of delaying touch events. Why is delay necessary? In most cases, delayMillis doesn't make much sense, as in scrollable scenarios, no one would hold down and wait for the page to become scrollable.

1. Why no use `interactionMode = UIKitInteropInteractionMode.NonCooperative`?
2. In most cases, a lightweight UIView just needs a tap event, and WKWebView is a heavyweight UIView, so it has its own scroll event. Scroll delay is the problem that scrolling gestures are not consumed by UIView in a long list of mixed rendered UIViews.

**Comment by elijah-semyonov** [comment]
@vickyleu 
No, you need to set `NonCooperative`. @Gaubee explained the rationale well

**Pull request merged**
